eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/MessageOnion.scala

```diff
@@ -83,31 +83,56 @@ object MessageOnion {
       if (records.get[InvoiceError].nonEmpty) return Left(ForbiddenTlv(UInt64(68)))
+      if (records.unknown.nonEmpty) return Left(ForbiddenTlv(records.unknown.head.tag))
       BlindedRouteData.validateMessageRelayData(blindedRecords).map(blindedRecords => IntermediatePayload(records, blindedRecords, nextBlinding))
     }
   }
 
   /** Per-hop payload for a final node. */
-  case class FinalPayload(records: TlvStream[OnionMessagePayloadTlv], blindedRecords: TlvStream[RouteBlindingEncryptedDataTlv]) extends PerHopPayload {
+  sealed trait FinalPayload extends PerHopPayload {
+    val blindedRecords: TlvStream[RouteBlindingEncryptedDataTlv]
     val pathId_opt: Option[ByteVector] = blindedRecords.get[RouteBlindingEncryptedDataTlv.PathId].map(_.data)
-    val replyPath_opt: Option[BlindedRoute] = records.get[ReplyPath].map(_.blindedRoute)
   }
 
+  case class InvoiceRequestPayload(records: TlvStream[OnionMessagePayloadTlv], blindedRecords: TlvStream[RouteBlindingEncryptedDataTlv]) extends FinalPayload {
+    val invoiceRequest: OfferTypes.InvoiceRequest = OfferTypes.InvoiceRequest(records.get[InvoiceRequest].get.tlvs)
+    val replyPath: BlindedRoute = records.get[ReplyPath].get.blindedRoute
+  }
+
+  case class InvoicePayload(records: TlvStream[OnionMessagePayloadTlv], blindedRecords: TlvStream[RouteBlindingEncryptedDataTlv]) extends FinalPayload {
+    val invoice: Bolt12Invoice = Bolt12Invoice(records.get[Invoice].get.tlvs)
+  }
+
+  case class InvoiceErrorPayload(records: TlvStream[OnionMessagePayloadTlv], blindedRecords: TlvStream[RouteBlindingEncryptedDataTlv]) extends FinalPayload {
+    val invoiceError: OfferTypes.InvoiceError = OfferTypes.InvoiceError(records.get[InvoiceError].get.tlvs)
+  }
+
+  /** This payload is invalid but the blinded records are valid. */
+  case class InvalidResponsePayload(records: TlvStream[OnionMessagePayloadTlv], blindedRecords: TlvStream[RouteBlindingEncryptedDataTlv], failure: InvalidTlvPayload) extends FinalPayload
+
   object FinalPayload {
     def validate(records: TlvStream[OnionMessagePayloadTlv], blindedRecords: TlvStream[RouteBlindingEncryptedDataTlv]): Either[InvalidTlvPayload, FinalPayload] = {
-      records.get[InvoiceRequest].map(i => OfferTypes.InvoiceRequest.validate(i.tlvs)) match {
-        case Some(Left(failure)) => return Left(failure)
-        case _ => // valid or missing
-      }
-      records.get[Invoice].map(i => Bolt12Invoice.validate(i.tlvs)) match {
-        case Some(Left(failure)) => return Left(failure)
-        case _ => // valid or missing
-      }
-      records.get[InvoiceError].map(i => OfferTypes.InvoiceError.validate(i.tlvs)) match {
-        case Some(Left(failure)) => return Left(failure)
-        case _ => // valid or missing
-      }
-      BlindedRouteData.validateMessageRecipientData(blindedRecords).map(blindedRecords => FinalPayload(records, blindedRecords))
+      BlindedRouteData.validateMessageRecipientData(blindedRecords).map(_ =>
```

Comment:
This deserves a few unit tests for the messages we reject that we didn't reject before (for example, an invoice with a reply path, or an invoice with an invoice error, etc). We should also verify that we correctly return the right type when receiving an `invoice_request` or an `invoice`.

Reply:
Done